### PR TITLE
Send RPC request to switch assets directory on hot reload.

### PR DESCRIPTION
**Pull request by aam**
This is needed to pick up updated assets that are expected to be picked up on hot reload.

This relies on https://github.com/flutter/engine/pull/4323

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -1314,6 +1314,16 @@ class FlutterView extends ServiceObject {
     await subscription.cancel();
   }
 
+  Future<Null> setAssetDirectory(Uri assetsDirectory) async {
+    if (assetsDirectory != null) {
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/vmservice.dart]: when will the assetsDirectory be null?
I am wondering if it is sufficient to assert that assetsDirectory is not null.

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -127,18 +127,27 @@ class FlutterDevice {
...
+        packagesUri: devicePackagesUri,
       );
       reports.add(report);
     }
     return reports;
   }
 
+  Future<Null> resetAssetDirectory() async {
+    final Uri deviceAssetsDirectoryUri = devFS.baseUri.resolveUri(
+        fs.path.toUri(getAssetBuildDirectory()));
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/resident_runner.dart]: assert deviceAssetsDirectoryUri != null ?

**Review by a-siva — APPROVED** [review]

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 2 above)
Comment (aam) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Done

File: packages/flutter_tools/lib/src/vmservice.dart (hunk 1 above)
Comment (aam) [reply packages/flutter_tools/lib/src/vmservice.dart]: Good point, done.

**Comment by Hixie** [comment]
cc @cbracken for review

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 3]
```diff
@@ -139,6 +139,15 @@ class FlutterDevice {
...
+    assert(deviceAssetsDirectoryUri != null);
+    for (FlutterView view in views) {
+      await view.setAssetDirectory(deviceAssetsDirectoryUri);
+    }
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/resident_runner.dart]: Maybe very slightly better?
```dart
await Future.wait(views.map((view) => view.setAssetDirectory(deviceAssetsDirectoryUri)));
```

**Review by cbracken — APPROVED** [review]

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 4]
```diff
@@ -140,6 +140,13 @@ class FlutterDevice {
     return reports;
   }
 
+  Future<Null> resetAssetDirectory() async {
+    final Uri deviceAssetsDirectoryUri = devFS.baseUri.resolveUri(
+        fs.path.toUri(getAssetBuildDirectory()));
+    assert(deviceAssetsDirectoryUri != null);
+    await Future.wait(views.map((view) => view.setAssetDirectory(deviceAssetsDirectoryUri)));
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/resident_runner.dart]: `FlutterView view`

**Review by cbracken — APPROVED** [review]

**Comment by cbracken** [comment]
lgtm

**Comment by aam** [comment]
Currently this is pending on https://github.com/flutter/engine/pull/4323 being reviewed and landed.

**Comment by Hixie** [comment]
This seems to have dramatically regressed some of the hot reload benchmarks.

**Comment by aam** [comment]
There is something fishy going on with those benchmarks. See discussion on original commit: https://github.com/flutter/flutter/commit/8da5af55c513511da75042c0a9d4beaf0597af5e#commitcomment-26676085

**Pull request merged**